internal/framework/fetch/fetch_test.go

```diff
@@ -0,0 +1,720 @@
+				testURL = server.URL + tt.url
+			}
+
+			data, err := testFetcher.GetRemoteFile(testURL, tt.options...)
+
+			if tt.expectErr {
+				if err == nil {
+					t.Error("Expected error, got nil")
+				}
+				if tt.expectedErr != "" && !strings.Contains(err.Error(), tt.expectedErr) {
+					t.Errorf("Expected error containing %q, got: %v", tt.expectedErr, err)
+				}
+			} else if err != nil {
+				t.Errorf("Unexpected error: %v", err)
+			}
+
+			if tt.validateFunc != nil {
+				tt.validateFunc(t, data, err)
+			}
+		})
+	}
+}
+
+func TestGetRemoteFileError(t *testing.T) {
+	fetcher, err := NewDefaultFetcher()
+	if err != nil {
+		t.Fatalf("NewDefaultFetcher() failed: %v", err)
+	}
+
+	tests := []struct {
+		setupServer   func() *httptest.Server
+		name          string
+		url           string
+		expectErrType string
+		options       []Option
+	}{
+		{
+			name: "HTTP error response",
+			setupServer: func() *httptest.Server {
+				return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+					w.WriteHeader(http.StatusNotFound)
+				}))
+			},
+			options:       []Option{WithRetryAttempts(0)},
+			expectErrType: "HTTPError",
+		},
+		{
+			name:          "network connection error",
+			url:           "http://127.0.0.1:1",
+			options:       []Option{WithRetryAttempts(0), WithTimeout(10 * time.Millisecond)},
+			expectErrType: "HTTPError",
+		},
+		{
+			name: "timeout during request",
+			setupServer: func() *httptest.Server {
+				return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+					time.Sleep(20 * time.Millisecond)
+					w.WriteHeader(http.StatusOK)
+					_, _ = w.Write([]byte("delayed response"))
+				}))
+			},
+			options:       []Option{WithTimeout(10 * time.Millisecond), WithRetryAttempts(0)},
+			expectErrType: "HTTPError",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testURL := tt.url
+			if tt.setupServer != nil {
+				server := tt.setupServer()
+				defer server.Close()
+				testURL = server.URL
+			}
+
+			_, err := fetcher.GetRemoteFile(testURL, tt.options...)
+			if err == nil {
+				t.Error("Expected error, got nil")
+			}
+
+			if tt.expectErrType != "" {
+				switch tt.expectErrType {
+				case "HTTPError":
+					var httpErr *HTTPError
+					if !errors.As(err, &httpErr) {
+						t.Errorf("Expected HTTPError, got %T: %v", err, err)
+					}
+				default:
+					t.Errorf("Unknown expected error type: %s", tt.expectErrType)
+				}
+			}
+		})
+	}
+}
+
+func TestParseS3URL(t *testing.T) {
+	tests := []struct {
+		name      string
+		url       string
+		bucket    string
+		key       string
+		expectErr bool
+	}{
+		// Error cases specific to parsing logic (not covered in integration tests)
+		{
+			name:      "missing key with trailing slash",
+			url:       "s3://bucket/",
+			expectErr: true,
+		},
+		{
+			name:      "invalid URL encoding",
+			url:       "s3://bucket/invalid%gg",
+			expectErr: true,
+		},
+	}
```

Comment:
IMO, with how simple this test is, I wouldn't even both with the test loop and just call the function directly with these two inputs.